cert/cert_v2.go

```diff
@@ -0,0 +1,655 @@
+package cert
+
+import (
+	"bytes"
+	"crypto/ecdh"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"net/netip"
+	"slices"
+	"time"
+
+	"golang.org/x/crypto/cryptobyte"
+	"golang.org/x/crypto/cryptobyte/asn1"
+	"golang.org/x/crypto/curve25519"
+)
+
+const (
+	classConstructed     = 0x20
+	classContextSpecific = 0x80
+
+	TagCertDetails   = 0 | classConstructed | classContextSpecific
+	TagCertCurve     = 1 | classContextSpecific
+	TagCertPublicKey = 2 | classContextSpecific
+	TagCertSignature = 3 | classContextSpecific
+
+	TagDetailsName           = 0 | classContextSpecific
+	TagDetailsNetworks       = 1 | classConstructed | classContextSpecific
+	TagDetailsUnsafeNetworks = 2 | classConstructed | classContextSpecific
+	TagDetailsGroups         = 3 | classConstructed | classContextSpecific
+	TagDetailsIsCA           = 4 | classContextSpecific
+	TagDetailsNotBefore      = 5 | classContextSpecific
+	TagDetailsNotAfter       = 6 | classContextSpecific
+	TagDetailsIssuer         = 7 | classContextSpecific
+)
+
+const (
+	// MaxCertificateSize is the maximum length a valid certificate can be
+	MaxCertificateSize = 65536
+
+	// MaxNameLength is limited to a maximum realistic DNS domain name to help facilitate DNS systems
+	MaxNameLength = 253
+
+	// MaxNetworkLength is the maximum length a network value can be.
+	// 16 bytes for an ipv6 address + 1 byte for the prefix length
+	MaxNetworkLength = 17
+)
+
+type certificateV2 struct {
+	details detailsV2
+
+	// RawDetails contains the entire asn.1 DER encoded Details struct
+	// This is to benefit forwards compatibility in signature checking.
+	// signature(RawDetails + Curve + PublicKey) == Signature
+	rawDetails []byte
+	curve      Curve
+	publicKey  []byte
+	signature  []byte
+}
+
+type detailsV2 struct {
+	name           string
+	networks       []netip.Prefix
+	unsafeNetworks []netip.Prefix
```

Comment:
I mentioned this as a reply to a resolved comment already, but in case it's folded and not seen: I think the name `networks` and `unsafeNetworks` are confusing. The former one is an address + its on-link network prefix, while the latter is just a network prefix. For the former, only a single address is routable to the host, while all addresses in the network is routable to the host for the latter.

I would propose the name to be `addresses` (I think it's fine even though this expects an CIDR, as the input is same as what `ip address add` command expects) which clearly indicate that the address part of the network prefix is significant, and `routable_networks` (or maybe `unsafe_routable_networks`) which clearly indicates that these are CIDRs for routing only.

Reply:
I share concerns of confusion with the naming here. I'm also comfortable sticking with the old name which at least creates no net new confusion: https://github.com/slackhq/nebula/pull/1216#discussion_r1759131003

Reply:
I'd also be open to the name `addresses` for the `networks` field. I think the name `networks` came up when we were considering it in the context of a CA certificate? Using that same field name for two very-similar-but-different purposes muddies the waters a little bit, but the intent was that it shows which _networks_ (as V2 certs let you have more than one!) the host participates in, as well as the actual address assignment on non-CA certs. However, the `ip address add` argument is very persuasive and probably much more obvious to new users.

Naming `unsafeNetworks` is a little trickier. When the `addresses` field is named `networks`, I think it's a pretty logical name, but `unsafeAddresses` doesn't really work. I'd propose something like `unsafeRouterFor`: the prefix `unsafe` links it mentally to `unsafe_routes` in the config, and I think `RouterFor` shows that the field expresses a capability, rather than something that might affect how "regular" Nebula traffic is routed.

Reply:
I think addresses places emphasis on the use of the ip; networks on the prefix use. But ultimately both are used and the content passed is CIDRs? Perhaps better to reflect what it is and what is passed in by the user rather than the things it’s used for (similar to before where it was called Ips). 

The bigger challenge before was having to manually change the ip when using ParseCIDR because it returned the base address instead of the provided IP, and the confusion of the certificate field ‘subnets’. Both of those are no longer an issue so bounds better.